quodlibet/browsers/covergrid/models.py

```diff
@@ -0,0 +1,254 @@
+            model.connect('rows-reordered', self._rows_reordered)
+        ]
+
+    def destroy(self):
+        model = self._model
+        for sig in self.__sigs:
+            model.disconnect(sig)
+        self._model = None
+
+    @GObject.Property
+    def include_item_all(self):
+        return self.__include_item_all
+
+    @include_item_all.setter
+    def include_item_all(self, value):
+        if self.__include_item_all == value:
+            return
+        self.__include_item_all = value
+        removed, added = (0, 1) if value else (1, 0)
+        self.items_changed(0, removed, added)
+
+    @GObject.Property
+    def filter(self):
+        return self.__filter
+
+    @filter.setter
+    def filter(self, value):
+        b = background_filter()
+        if b is None and value is None:
+            f = None
+        elif b is None:
+            f = lambda item: value(item.album)
+        else:
+            f = lambda item: b(item.album) and value(item.album)
+        if f or self.__filter:
+            self.__filter = f
+            self._model.refilter()
+
+    def do_get_n_items(self):
+        return len(self)
+
+    def do_get_item(self, index):
+        return self[index]
+
+    def do_get_item_type(self):
+        return AlbumListItem
+
+    def __len__(self):
+        n = len(self._model)
+        if self.__include_item_all or n < 1:
+            return n
+        return n - 1
+
+    def __getitem__(self, index):
+        if not self.__include_item_all:
+            index += 1
+        return self._get_item(index)
+
+    def _get_item(self, index):
+        model = self._model
+        iter = model.iter_nth_child(None, index)
+        return model.get_value(iter) if iter else None
+
+    def _update_n_albums(self):
+        self.__item_all.props.n_albums = len(self._model) - 1
+
+    def _apply_filter(self, model, iter, _):
+        filter = self.__filter
+        if filter is None:
+            return True
+        item = model.get_value(iter)
+        if item is self.__item_all:
+            return True
+        return filter(item)
+
+    def _row_changed(self, model, path, iter):
+        item = model.get_value(iter)
+        item.notify('album')
+
+    def _row_inserted(self, model, path, iter):
+        # Ensure that the tree model will emit the "row-changed" signal
+        model.ref_node(iter)
+
+        index = path.get_indices()[0]
+        if not self.__include_item_all:
+            index -= 1
+        if index >= 0:
+            self.items_changed(index, 0, 1)
+            self._update_n_albums()
+
+    def _row_deleted(self, model, path):
+        index = path.get_indices()[0]
+        if not self.__include_item_all:
+            index -= 1
+        if index >= 0:
+            self.items_changed(index, 1, 0)
+            self._update_n_albums()
+
+    def _rows_reordered(self, model, path, iter, new_order):
+        n = len(self)
+        self.items_changed(0, n, n)
+
+
+class AlbumListSortModel(ObjectModelSort):
```

Comment:
Why do we need this? Perhaps you can add a docstring here...

Reply:
We don't _need_ it. It's an indirection to make `models.py` the single place to change implementation of the models.